### ISSUE TITLE
Lagging on playing on small MinnowBoard

### ISSUE THREAD

**Issue by IntelligyScience** [caprica/vlcj#406]
We have a JavaFX (scalaFX) Project. I build the player with the demo of JavaFX.

On Mac the player runs without lagging. On the minoowboard it lags. 
Is their a way to improve this?

Thanks

**Comment by caprica** [comment]
I don't know what "minoowboard" is, but I'm guessing it's something low spec onto which you're trying to run a JVM, Scala, JavaFX and do software rendering. I'm not surprised you see lag.

All you can do with the JavaFX approach is either reduce the amount of video data you request from VLC (be it resolution or chroma format), then optimise your JavaFX rendering of the video data. Typically this means making sure you are using hardware-accelerated surfaces, assuming your "minoowboard" can provide hardware acceleration.

Failing that, run a profiler and find out where the lag is.

There's no magic bullet to solve performance problems. It's lagging because it's not quick enough, sorry for stating the obvious!

**Comment by IntelligyScience** [comment]
Yes this was my opinion too. But if I play the video (H264 with HD) with the vlc, it plays without any lag.
That makes me nervous. 

here is a few example code
<code>
class CanvasPlayerComponent(wImage: WritableImage) extends DirectMediaPlayerComponent(new CanvasBufferFormatCallback()) {


    private[gui] var pixelWriter: PixelWriter = null

    private def getPW: PixelWriter = {
      if (pixelWriter == null) {
        pixelWriter = writableImage.getPixelWriter
      }
      pixelWriter
    }

    override def finished(mediaPlayer: MediaPlayer) {
      super.finished(mediaPlayer)
      mediaPlayer.stop()
    }

    override def stopped(mediaPlayer: MediaPlayer) {
      super.stopped(mediaPlayer)
      System.gc()
      System.runFinalization()
      Try( mediaPlayerBuffer.remove(1) ) match {
        case Success(elem) =>
          Platform.runLater{
            mediaPlayer.prepareMedia(elem.completePath(manager.contentPath))
            mediaPlayer.start()
            manager.logDisplay(elem.name)
          }
        case _ => Platform.runLater(loadContent())
      }
    }

    override def display(mediaPlayer: DirectMediaPlayer, nativeBuffers: Array[Memory], bufferFormat: BufferFormat) {
      if (wImage == null) {
        return
      }
      Platform.runLater{
        val nativeBuffer = mediaPlayer.lock().head
        try {
          val byteBuffer = nativeBuffer.getByteBuffer(0, nativeBuffer.size())
          getPW.setPixels(0, 0, bufferFormat.getWidth, bufferFormat.getHeight, pixelFormat, byteBuffer, bufferFormat.getPitches.head )
        } finally {
          mediaPlayer.unlock()
        }
        ()
      }
    }
  }

  class CanvasBufferFormatCallback extends BufferFormatCallback {
    def getBufferFormat(sourceWidth: Int, sourceHeight: Int): BufferFormat = {
      //val visualBounds: Rectangle2D = Screen.primary.getVisualBounds
      /*Platform.runLater{
        videoSourceRatioProperty.set(sourceHeight.toFloat / sourceWidth.toFloat)
      }*/
      new RV32BufferFormat(/*visualBounds.getWidth.toInt, visualBounds.getHeight.toInt*/maxX.toInt,maxY.toInt)
    }
  }
</code>

**Comment by caprica** [comment]
It plays without lag in VLC itself because VLC has a native video output for that platform. This would be same performance as using a vlcj "embedded" media player in an AWT Canvas rather than the direct rendering approach.

The direct rendering approach is basically you rendering the video in software. If you work the maths out for the screen resolution, the bit-depth and the FPS you want, you can work out how much work you're trying to do. The direct rendering approach is CPU intensive.

Bottom line is you're trying to play full HD on a low-powered device, in software.

**Comment by IntelligyScience** [comment]
Yeah. That is the problem. But this problem comes from my customer. He want this in this way.

What you mean. Is it better to use a swing GUI and use vlcj in Swing/AWT?
Do you think this could improve the performance?

**Comment by caprica** [comment]
If you can use an AWT Canvas, the performance will be better. It should be on a par with native performance, but there will be some overhead because of the JVM, and Scala in your case.

**Comment by caprica** [comment]
It would be even better to implement it natively in C/C++ of course... :-D

**Comment by IntelligyScience** [comment]
yeah C/C++ would be the best. But they need it OS free :(

What you mean about overhead. Only Code or even memory overhead which ends in a crash?

**Comment by caprica** [comment]
The idea with the embedded media player is that you tell VLC to render its video into a component of your choosing. VLC renders the video, so it should be as fast as native performance. There will inevitably be a small amount of overhead because you are running inside a JVM, I don't know how much overhead there will be, but there will be some. Adding Scala adds more code doesn't it, so that will have another small amount of overhead. But probably this overhead won't amount to much.

You need to make sure hardware acceleration works.

So I'm talking about performance overhead, nothing else. Memory usage should be stable if you do things the right way.

**Comment by IntelligyScience** [comment]
Okay, that sounds good. Can you tell me, and I hop so, how I have to tell vlc to render this video?

**Comment by caprica** [comment]
Look at the Javadoc, the vlcj test sources, or my company web-site tutorials for Embedded Media Player:

http://capricasoftware.co.uk/#/projects/vlcj/tutorial

There's also a full example:

https://github.com/caprica/vlcj-player
